### DISCUSSION TITLE
Possíveis features para o plugin

### DISCUSSION THREAD

**Discussion by AssisRaphael** [pmartinsesa/blip-addons#7] [Ideas]
- Permitir o usuário alterar a cor do bloco. Ao clicar com o dedo direito no bloco aparecer a opção "Alterar cor"
- Permitir o usuário mudar o formato do bloco (redondo, elipse, quadrado, etc)
- Adicionar actions aos blocos no Blip para realizar process comands com uma interface gráfica
- Adicionar funcionalidade para verificar se existem erros no fluxo (Ex: limite de 35 blocos sem inputs, verificar se existe uma tracking que pode ser nula, procurar por loops)
- Reposicionar todos os blocos para organizar o fluxo
- Adicionar tags automáticas nos blocos
- Adicionar as funções de colocar em negrito, itálico, tarjado e sublinhado os textos dentro de conteúdos nos blocos

Commands a serem desenvolvidas:

- Enviar email
- Enviar uma mensagem agendada
- Obter informações sobre atendentes online
- Enviar relatório de atendimento por email

**Comment by AssisRaphael** [comment]
## Adicionar novas actions no blip

### Ideia geral

A Ideia é ter novas actions na opção de actions de um bloco. A primeira action extra para desenvolver é a de envio de email.

### Analise da solução

Devido à limitação do Blip, não conseguimos adicionar novas actions, pois o blip necessariamente irá verificar se a action é uma dos tipos predefinidos. Também não é possível duplicar uma action de mesmo tipo. Por conta disso, a ideia para desenvolver esta feature é a seguinte:

1. Alterar o nome da action "ProcessCommand" para "Outros" ou "Extras", algo assim.
2. Dentro da tela de ProcessCommand, personalizar a tela para adicionar as novas actions.
3. Adicionar uma seleção de lista com as actions extras adicionadas e exibir a tela personalizada de cada

**Reply by gabrielpetrone** [reply]
Sensacional! Seria massa pensar numa forma de escalar essa feature, de modo que eu consiga adicionar uma nova Action. A Action inclusive poderia ser um pacote de ações (HTTP Requests + Scripts) para já entregar tudo pronto pros clientes. Exemplo o Marketplace do CloneBots.

Aplicação:
A solução de ["Informar a posição na Fila"](https://forum.blip.ai/t/informe-a-posicao-de-fila-e-tempo-medio-de-espera-para-o-seu-cliente/10947) poderia ser adicionada com um único clique.

**Reply by AssisRaphael** [reply]
Sobre essa ideia de adicionar um pacote de ações com um clique, acho que será possível quando for lançada a feature de subfluxos no Blip. A gente poderia simplesmente inserir o subfluxo escolhido no bot. Enquanto essa feature não lança, a gente poderia inserir os blocos mesmo no fluxo. Estou pensando nos cenários possíveis, como:

* O pacote de ações está contido em um único bloco
* O pacote de ações está contido em N blocos

Para o primeiro caso seria bem de boa, mas quando o pacote começa a envolver vários blocos pode ficar um pouco bagunçado, e para esse caso eu acho que a solução com os subfluxos seria melhor. Outro ponto é como cadastraríamos esse pacote de ações. Seria algo fixo no plugin (somente alguns pacotes disponíveis)? Existiria algum lugar para os usuários acrescentarem seus pacotes de ações para outros usuários usarem? Se sim, como fazer isso funcionar no plugin?

Acho que se a plataforma adicionar um marketplace de pacotes de ações daria. Assim os usuários cadastrariam seus pacotes de ações lá e o plugin acessaria.

**Comment by gabrielpetrone** [comment]
# Importar e Exportar Snnipets.

## Ideia geral
Possibilidade de salvar e compartilhar códigos snippets com os colegas.

**Reply by AssisRaphael** [reply]
Isso já esta no nosso roadmap. Na próxima versão já iremos incluir.

**Reply by gabrielpetrone** [reply]
Top @AssisRaphael , enquanto não lança a feature, sabe me dizer se é possível compartilhar os snnipets copiando algum arquivo do cache do Google Chrome? Procurei aqui mas não achei o arquivo que salvou meu código.

**Reply by AssisRaphael** [reply]
No momento você teria que enviar para a outra pessoa o código e o nome de cada snippet para ela adicionar no plugin dela. Na tela de configurações dos snippets tem todas essas informações. Ai teria que ser passado um por um e a pessoa teria adicionar um por um também.

![image](https://user-images.githubusercontent.com/58819530/165835017-710d176b-a8e1-49d6-bb3d-566fb6937300.png)

Se você enviar as suas configurações direto, pegando na cache do navegador, você sobrescreveria todos os snippets da outra pessoa. E também nem sei se daria pra passar assim, porque salvamos direto na conta Google do usuário.

**Comment by AssisRaphael** [comment]
Seria interessante adicionar a funcionalidade de alterar a cor e/ou o formato do bloco automaticamente de acordo com algumas regras personalizáveis. Por exemplo:

* Se um bloco tem mais de 2 condições de saída
* Se o bloco não tiver nenhuma condição de saída (somente a saída default)
* Se o nome do bloco contém uma determinada string, ex: "[SELECT]", "[SET]"
* Se o bloco possui somente uma certa action, ex: "redirect", "setVariable", "sendMessage"

Neste caso, a configuração disso poderia ficar da seguinte maneira:

obs: o título seria configurações da edição de blocos automática (algo assim)
![image](https://user-images.githubusercontent.com/58819530/165636775-a5f2d24e-ebf7-42c2-b802-fc38dcb1de7a.png)

E ao expandir cada configuração a tela seria igual a de edição dos blocos:
![image](https://user-images.githubusercontent.com/58819530/165636990-53ae174f-4ca7-4839-9b21-ae47e68d647e.png)

**Comment by AssisRaphael** [comment]
* Poder selecionar vários blocos e editar todos eles de uma única vez (BETA)
* Adicionar na verificação de inconsistências o bug da condição "variável diferente de" no blip que executa um "ou" ao invés de um "e". Assim, caso o usuário coloque mais de uma opção dentro dessa validação a gente reclama. (BETA)
![image](https://user-images.githubusercontent.com/58819530/171035638-422f8020-f145-431b-b0fc-a16cde598702.png)
* Realizar o merge de blocos. Em um processo de refatoração de um fluxo eu posso querer unificar dois blocos e seria interessante que o plugin fizesse isso. A princípio, imagino que poderíamos selecionar os N blocos que queremos unificar e ao selecionar uma opção (pode ser um botão) o plugin geraria um novo bloco com todas as ações dos blocos selecionados, incluindo mensagens a serem exibidas e condições de saída. Como uma evolução dessa ideia, quando o usuário selecionar mais de um bloco, podemos mostrar um menu com várias opções como esse merge, a edição desses N blocos, reposicionar eles para uma determinada geometria, etc.
* Permitir copiar e colar ações de um bloco para outro.
* Adicionar o atalho `ctrl+ f` que abre a barra de busca. Ao pesquisar um termo, abaixo da barra de pesquisa aparece uma lista de todos os blocos que deram match com aquele termo pesquisado. Ao clicar em um dos blocos da lista a interface foca no bloco selecionado.

**Comment by gabrielpetrone** [comment]
Parace bobo/inútil mas poderia ter um atalho para Duplicar um Bloco (**CTRL + D**) ao invés de ter que copiar e colar (**CTRL + C & CTRL + V)**.

**Comment by AssisRaphael** [comment]
- Permitir o usuário alterar o nome de uma variável que ele criou no fluxo. Para isso a alteração precisaria ser feita em todos os pontos daquele bot (ações, condições de saída, condições de execução, textos com {{variavel}}, etc ).

**Comment by AssisRaphael** [comment]
Permitir adicionar comentários no bot. Seria interessante ter uma funcionalidade de comentários no Blip similar ao que tem no figma. Em alguns momentos é útil esclarecer algumas lógicas no bot que se referem a regras de negócio para facilitar a manutenção do fluxo.
